### ISSUE TITLE
false positive for `if-let-rescope` where `if let` statements without `else` are also linted

### ISSUE THREAD

**Issue by m4rch3n1ng** [rust-lang/rust#133167]
### Code

remove the `else` from the example at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/if_let_rescope/static.IF_LET_RESCOPE.html

```rust
#![warn(if_let_rescope)]
#![allow(unused_variables)]

struct Droppy;
impl Drop for Droppy {
	fn drop(&mut self) {
		// Custom destructor, including this `drop` implementation, is considered
		// significant.
		// Rust does not check whether this destructor emits side-effects that can
		// lead to observable change in program semantics, when the drop order changes.
		// Rust biases to be on the safe side, so that you can apply discretion whether
		// this change indeed breaches any contract or specification that your code needs
		// to honour.
		println!("dropped");
	}
}
impl Droppy {
	fn get(&self) -> Option<u8> {
		None
	}
}

fn main() {
	if let Some(value) = Droppy.get() {
		// do something
	}
}
```

### Current output

```text
warning: `if let` assigns a shorter lifetime since Edition 2024
  --> src/main.rs:24:5
   |
24 |     if let Some(value) = Droppy.get() {
   |        ^^^^^^^^^^^^^^^^^^------^^^^^^
   |                          |
   |                          this value has a significant drop implementation which may observe a major change in drop order and requires your discretion
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see issue #124085 <https://github.com/rust-lang/rust/issues/124085>
help: the value is now dropped here in Edition 2024
  --> src/main.rs:26:2
   |
26 |     }
   |     ^
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(if_let_rescope)]
   |         ^^^^^^^^^^^^^^
help: a `match` with a single arm can preserve the drop order up to Edition 2021
   |
24 ~     match Droppy.get() { Some(value) => {
25 |         // do something
26 ~     } _ => {}}
   |
```

### Desired output

```text
```

### Rationale and extra context

as far as i know, the new if-let-rescope is only relevant, when you have an `else` block, like in the actual [example](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/if_let_rescope/static.IF_LET_RESCOPE.html), because it cannot change to be dropped before the `else`, as there is no else, so (as far as i know) the place where it is dropped does not change.

<details>

<summary>more info</summary>

if you change the code to print stuff

```rust
#![warn(if_let_rescope)]
#![allow(unused_variables)]

struct Droppy;
impl Drop for Droppy {
	fn drop(&mut self) {
		dbg!("drop");
	}
}
impl Droppy {
	fn get(&self) -> Option<u8> {
		Some(0)
	}
}

fn main() {
	if let Some(value) = Droppy.get() {
		dbg!("one");
	}
	dbg!("two");
}
```

and then run it with both edition 2024 and edition 2021

```text
$ rustup run nightly -- rustc --edition 2024 -Zunstable-options src/main.rs 
$ ./main
[src/main.rs:18:3] "one" = "one"
[src/main.rs:7:3] "drop" = "drop"
[src/main.rs:20:2] "two" = "two"
$ rustup run nightly -- rustc --edition 2021 src/main.rs 
$ ./main
[src/main.rs:18:3] "one" = "one"
[src/main.rs:7:3] "drop" = "drop"
[src/main.rs:20:2] "two" = "two"
```

they both do the same

</details>

### Other cases

while testing i found out that if you modify the original code to include anything after the `if let` statement  it doesn't lint it anymore

```rust
#![warn(if_let_rescope)]
#![allow(unused_variables)]

struct Droppy;
impl Drop for Droppy {
	fn drop(&mut self) {
		println!("dropped");
	}
}
impl Droppy {
	fn get(&self) -> Option<u8> {
		None
	}
}

fn main() {
	if let Some(value) = Droppy.get() {
		// do something
	}
	// this code is still affected by the lint
}
```

```rust
#![warn(if_let_rescope)]
#![allow(unused_variables)]

struct Droppy;
impl Drop for Droppy {
	fn drop(&mut self) {
		println!("dropped");
	}
}
impl Droppy {
	fn get(&self) -> Option<u8> {
		None
	}
}

fn main() {
	if let Some(value) = Droppy.get() {
		// do something
	}
	println!("with this line it doesn't trigger the lint anymore");
}
```

### Rust Version

```text
rustc 1.84.0-nightly (5ec7d6eee 2024-11-17)
binary: rustc
commit-hash: 5ec7d6eee7e0f5236ec1559499070eaf836bc608
commit-date: 2024-11-17
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.3
```

**Comment by jieyouxu** [comment]
cc @dingxiangfei2009 as you may know more about the intention.

**Comment by jieyouxu** [comment]
Marking this as `C-discussion` for now but please change it `C-bug` if more suitable.

**Comment by Turbo87** [comment]
FWIW I've just followed https://blog.rust-lang.org/2024/11/27/Rust-2024-public-testing.html to test out the Edition 2024 on the https://github.com/rust-lang/crates.io/ codebase and it looks like literally all of the suggested fixes were false positives. if I run `cargo +nightly check` without those fixes applied it passes without any warnings. luckily crates.io is a relatively small codebase, but for larger projects this could cause quite a bit of busy work.

**Comment by traviscross** [comment]
We talked about this extensively in the edition call yesterday.  Note that, while in simple cases like this it is a false positive, there are cases where the behavior does change for `if let` expressions without `else` branches due to the fact that the temporaries now do not live throughout the statement.

**Comment by jieyouxu** [comment]
> We talked about this extensively in the edition call yesterday. Note that, while in simple cases like this it is a false positive, there are cases where the behavior does change for `if let` expressions without `else` branches due to the fact that the temporaries now do not live throughout the statement.

If the distinction between such cases is not tractable to get right, maybe add a note for no-`else` cases that indicates the reason you explained, and that it requires additional attention?

**Comment by Turbo87** [comment]
IMHO false-negatives in this case seem better than this large number of false-positives. if there are false-negatives, wouldn't the compiler complain with a helpful message anyway?

**Comment by m4rch3n1ng** [comment]
fwiw, if i run `cargo +nightly fix --edition` in one of my projects, then clippy will give me 28 warnings for [`"single match"`](https://rust-lang.github.io/rust-clippy/master/index.html#single_match).

and also, the lint is only triggered if the `if let` statement is the last statement of the block that the `if let` was initiated in, if i put anything after the closing brackets (like i showed in the original post) the lint is not triggered at all, even without it affecting the `if let` at all (unless i'm misunderstanding the comments here).

so i think the best solution is to either lint every single `if let` statement (which would make the output extremely noisy and immediately result in a whole lot of clippy warnings) or to make this lint less aggressive (which might then result in a few runtime errors).

**Comment by Aloso** [comment]
Here's an example I got (minimized):

```rs
fn foo(mut iter: impl Iterator<Item = i32>) {
    if let Some(item) = iter.next() {
        _ = item;
    }
}
```

It says that `iter` has a significant drop implementation which may observe a major change in drop order.

This is obviously wrong: `iter` is not a temporary, it gets dropped at the end of the function in all editions.

But when I tried to verify, the warning went away when I inserted a `println!` statement after the `if let` statement. It only warns in this instance when the `if let` is the _last statement_ within the function or block 🤨

**Comment by mitsuhiko** [comment]
Is there a way to opt out a type from having a significant `Drop` implementation?  I just ran this through minijinja and from what I can tell every single `if let` that it has changed into a `match` is a false positive.

**Comment by vi** [comment]
Maybe there should be an option to explicitly exclude the `if let` migration, like `cargo fix --edition --except-of-if-let-to-match`, so it would handle things like `rng.gen()` to `r#gen`, but avoid the `match` patchbomb.

**Comment by rnestler** [comment]
A workaround is to run `cargo clippy --fix --allow-dirty` after running `cargo fix --edition` to revert the `if let` migration.

**Comment by mitsuhiko** [comment]
I ran this on a few more projects now and the false positive rate is 100% from what I can tell. I really think it would be good for projects to be able to opt out somehow from this migration lint.

**Comment by traviscross** [comment]
We're hopeful this will be improved by #133753 which is in the process of being merged.

**Comment by traviscross** [comment]
Now that #133753 is merged, the examples here, and others that we've tried, such as in #133605, no longer produce false positives, so we'll go ahead and close this.

**Comment by epage** [comment]
I am, and it sounds like a lot of other people, are still seeing significant false positives.  I reverted every single `if let` change in cargo-edit (killercup/cargo-edit#923) and cargo-release (crate-ci/cargo-release#866, some I instead changed to let-else).

**Comment by steffahn** [comment]
Coming from [this new forum thread](https://users.rust-lang.org/t/rust-2024-compatibility-lint-and-the-if-let-temporary-scope/125969), my takeaway of remaining false positives is:

* if the `if let` matches on an *enum* and only *some* enum variant(s) have fields with potentially-significant destructor, but those are exhaustively matched in the non-`else` case of the `if let`, then there can be no actual change in behavior, because the `else` case is only reached in case the drop is *not* significant

* in the case of `cargo-edit`, this was `if let Some(…irrefutable…) = expr` on an `Option<T>`
* in the urlo thread, I’ve identified the case of `if let Err(…irrefutable…)` on a `Result<(), E>`

**Comment by steffahn** [comment]
I’ll open a separate issue for that:
* https://github.com/rust-lang/rust/issues/137376

**Comment by DavidAntliff** [comment]
Not sure if helpful, but 2021 to 2024 migration turned this (9 lines):

```rust
    let mut encoded = if let Ok(channel) = container.inner().get_channel(&channel_id) {
        if let Some(sequence) = channel.get_sequence(&seq_id) {
            emitter.encode(sequence.get_subsequences())?
        } else {
            vec![]
        }
    } else {
        vec![]
    };
```

Into (12 lines):

```rust
    let mut encoded = match container.inner().get_channel(&channel_id) {
        Ok(channel) => {
            if let Some(sequence) = channel.get_sequence(&seq_id) {
                emitter.encode(sequence.get_subsequences())?
            } else {
                vec![]
            }
        }
        _ => {
            vec![]
        }
    };
```

Is this working as intended? Not a huge fan, myself, but will defer to the tools...
